airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/factory.py

```diff
@@ -0,0 +1,68 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+from __future__ import annotations
+
+import copy
+import importlib
+from typing import Any, Mapping
+
+from airbyte_cdk.sources.declarative.create_partial import create
+from airbyte_cdk.sources.declarative.interpolation.jinja import JinjaInterpolation
+from airbyte_cdk.sources.declarative.types import Config
+
+
+class DeclarativeComponentFactory:
+    def __init__(self):
+        self._interpolator = JinjaInterpolation()
+
+    def create_component(self, component_definition: Mapping[str, Any], config: Config):
+        """
+
+        :param component_definition: mapping defining the object to create. It should have at least one field: `class_name`
+        :param config: Connector's config
+        :return: the object to create
+        """
+        kwargs = copy.deepcopy(component_definition)
+        class_name = kwargs.pop("class_name")
+        return self.build(class_name, config, **kwargs)
+
+    def build(self, class_name: str, config, **kwargs):
+        fqcn = class_name
+        split = fqcn.split(".")
+        module = ".".join(split[:-1])
+        class_name = split[-1]
+
+        # create components in kwargs before propagating them
+        if "kwargs" in kwargs:
```

Comment:
we want to create instantiate the content of the kwargs dict before creating the subcomponents so we can share values from kwargs.

I've used this to avoid redefining values ([mainly the parent stream in this case)](https://github.com/airbytehq/airbyte/pull/13129/files#diff-ab24c17b57a8072cd8a943f176c6594a05adc1324474e44a379dcda768c1796dR73-R76)

You could make a case that this is just extra magic and that we'd be better off retyping it

```
invoice_line_items_stream:
  ref: "*ref(partial_stream)"
  name: "invoice_line_items"
  primary_key: "id"
  retriever:
    ref: "*ref(substream_retriever)"
    parent_stream: "*ref(invoices_stream)"
<...>
    stream_slicer:
      ref: "*ref(substream_slicer)"
      parent_stream: "*ref(invoices_stream)"
```

any thoughts either way @brianjlai ?

Reply:
Ohhhh okay that makes more sense now. I wasn't really thinking about it in terms of kwargs as a sort of reserved block in the DSL to encapsulate values. That clears up a lot thanks!

I guess my main question is what's the benefit to encapsulating everything in kwargs instead of directly? I can see some benefits to kwargs so users don't have to worry about the order of their fields which would be a painful gotcha.

I think my only reservation is that kwargs is a very python concept and less language agnostic. That being said it impossible to fully be agnostic since we'll always have things like classes so I'm not super opinionated about this and how it is now is fine too. I do remember in planning phase there was the concept of `vars` at one point and maybe that is a good way to keep the name generic.

Reply:
Agreed. Let's rename it to something other than vars because it's also a python builtin function. what do you think of `options` or `params`?